packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -581,24 +609,33 @@ export class ApolloServerBase {
     if (this.requestOptions.persistedQueries?.cache) {
       service.persistedQueries = {
         cache: this.requestOptions.persistedQueries.cache,
-      }
+      };
     }
 
-    await Promise.all(
-      this.plugins.map(
-        plugin =>
-          plugin.serverWillStart &&
-          plugin.serverWillStart(service),
-      ),
+    const serverListeners = (
+      await Promise.all(
+        this.plugins.map(
+          (plugin) => plugin.serverWillStart && plugin.serverWillStart(service),
+        ),
+      )
+    ).filter(
+      (maybeServerListener): maybeServerListener is GraphQLServerListener =>
+        typeof maybeServerListener === 'object' &&
+        !!maybeServerListener.serverWillStop,
     );
+    this.toDispose.add(async () => {
+      await Promise.all(
+        serverListeners.map(({ serverWillStop }) => serverWillStop?.()),
+      );
+    });
```

Comment:
I think the additional `async/await` here is just extraneous (though harmless)

```suggestion
    this.toDispose.add(() => {
      return Promise.all(
        serverListeners.map(({ serverWillStop }) => serverWillStop?.()),
      );
    });
```

Reply:
Your suggestion here doesn't typecheck because toDispose is supposed to return `ValueOrPromise<void>` but when I return Promise.all it ends up with `ValueOrPromise<(void | undefined)[]>`.  (The code is semantically different as I had not written `return await`.) I guess we could change the typing on toDispose, or I could throw in a `.then(() => {})` or something, but the async/await version seems clearer to me than either of those. Let me know if you disagree!